src/core/dbus-service.c

```diff
@@ -718,6 +747,54 @@ static int bus_service_set_transient_property(
         if (streq(name, "ReloadSignal"))
                 return bus_set_transient_reload_signal(u, name, &s->reload_signal, message, flags, error);
 
+        if (streq(name, "ExtraFileDescriptors")) {
+                int fd;
+                const char *fdname;
+
+                r = sd_bus_message_enter_container(message, 'a', "(hs)");
+                if (r < 0)
+                        return r;
+
+                while ((r = sd_bus_message_read(message, "(hs)", &fd, &fdname)) > 0) {
+                        _cleanup_(service_extra_fd_freep) ServiceExtraFD *efd = NULL;
+
+                        /* Disallow empty string for ExtraFileDescriptors.
+                         * Unlike OpenFile, StandardInput and friends, there isn't a good sane
+                         * default for an arbitrary FD. */
+                        if (fd < 0 || !fdname_is_valid(fdname) || isempty(fdname))
+                                return -EINVAL;
+
+                        if (UNIT_WRITE_FLAGS_NOOP(flags))
+                                continue;
+
+                        efd = new(ServiceExtraFD, 1);
+                        if (!efd)
+                                return -ENOMEM;
+
+                        *efd = (ServiceExtraFD) {
+                                .fd = -EBADF,
+                                .fdname = strdup(fdname),
+                        };
+
+                        if (!efd->fdname)
+                                return -ENOMEM;
+
+                        efd->fd = fcntl(fd, F_DUPFD_CLOEXEC, 3);
+                        if (efd->fd < 0)
+                                return -errno;
+
+                        LIST_APPEND(extra_fd, s->extra_fds, TAKE_PTR(efd));
+                }
+                if (r < 0)
+                        return r;
```

Comment:
Hm, this is slightly late after `r = sd_bus_message_read(message, "(hs)", &fd, &fdname)`. Please do something like the following:
```c
for (;;) {
        r = sd_bus_message_read(message, "(hs)", &fd, &fdname);
        if (r < 0)
                return r;
        if (r == 0)
                break;

        ....
}
```